Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -50,39 +47,283 @@ extension Workspace {
+            sharedMirrorFile: AbsolutePath?,
+            fileSystem: FileSystem
+        ) throws {
+            let localMirrorConfigFile = Workspace.DefaultLocations.mirrorsConfigurationFile(forRootPackage: rootPath)
+            try self.init(
+                localMirrorFile: localMirrorConfigFile,
+                sharedMirrorFile: sharedMirrorFile,
+                fileSystem: fileSystem
+            )
+        }
+
+        /// Initialize the workspace mirrors configuration
+        ///
+        /// - Parameters:
+        ///   - localMirrorFile: Path to the workspace mirrors configuration file
+        ///   - sharedMirrorFile: Path to the shared mirrors configuration file, defaults to the standard location.
+        ///   - fileSystem: The file system to use.
+        public init(
+            localMirrorFile: AbsolutePath,
+            sharedMirrorFile: AbsolutePath?,
+            fileSystem: FileSystem
+        ) throws {
+            self.localMirrors = .init(path: localMirrorFile, fileSystem: fileSystem, deleteWhenEmpty: true)
+            self.sharedMirrors = sharedMirrorFile.map { .init(path: $0, fileSystem: fileSystem, deleteWhenEmpty: false) }
+            self.fileSystem = fileSystem
+            // computes the initial mirrors
+            self._mirrors = DependencyMirrors()
+            try self.computeMirrors()
+        }
+
+        @discardableResult
+        public func applyLocal(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            try self.localMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        @discardableResult
+        public func applyShared(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            guard let sharedMirrors = self.sharedMirrors else {
+                throw InternalError("shared mirrors not configured")
+            }
+            try sharedMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        // 👀 is this correct, or should we merge?
+        // mutating the state we hold since we are passing it by reference to the workspace
+        // access should be done using a lock
+        private func computeMirrors() throws {
+            try self.lock.withLock {
+                self._mirrors.removeAll()
+
+                // prefer local mirrors to shared ones
+                let local = try self.localMirrors.mirrors()
+                if !local.isEmpty {
+                    self._mirrors.append(contentsOf: local)
+                    return
+                }
+
+                // use shared if local was not found or empty
+                let shared = try self.sharedMirrors?.mirrors() ?? [:]
+                if !shared.isEmpty {
+                    self._mirrors.append(contentsOf: shared)
+                }
+            }
+        }
+    }
+
+    public struct Mirrors {
```

Comment:
What do you think of renaming this to `MirrorsConfigurationFile`, or something to emphasize the file interaction?

Reply:
this is currently `Configuration.Mirrors` so maybe `Configuration.MirrorsFile`?

Reply:
went for MirrorsStorage